# Importing Libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing Data Bases

In [60]:
# Import orders.csv
# Create a string of the path
# Define the path variable
path=r'/Users/douniaelyoussoufi/Achievement 4 '
# Load the Orders CSV file directly using os.path.join
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col=False)
# Import orders_products_prior
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col=False)

# Step 1: Exercise 4.6

In [25]:
# Check the dataframe of df_ords_prior
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [10]:
# Check the dimensions of df_ords_prior
df_ords_prior.shape

(32434489, 4)

In [27]:
# Check the dimensions of df_ords
df_ords.shape

(3421083, 7)

In [23]:
# Merge df_ords_prior with df_ords using the indicator argument
df_merge = df_ords_prior.merge(df_ords, on='order_id', indicator=True)
# Display the first five rows to check the merge results
df_merge.head()

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0         2       33120                  1          1   202279    prior   
1         2       28985                  2          1   202279    prior   
2         2        9327                  3          0   202279    prior   
3         2       45918                  4          1   202279    prior   
4         2       30035                  5          0   202279    prior   

   order_number  order_dow  order_hour_of_day  days_since_prior_order _merge  
0             3          5                  9                     8.0   both  
1             3          5                  9                     8.0   both  
2             3          5                  9                     8.0   both  
3             3          5                  9                     8.0   both  
4             3          5                  9                     8.0   both

In [33]:
# Check the value counts of the _merge column
df_merge['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

Comment: This indicates a perfect full match between df_ords_prior and df_ords. Every order_id in both datasets overlaps entirely, with no rows exclusive to either dataset

In [35]:
# Merge df_ords_prior with df_ords using the argument how = 'outer'
df_merged = df_ords_prior.merge(df_ords, on='order_id', how='outer', indicator=True)
# Check the value counts of the _merge column
df_merged['_merge'].value_counts()

_merge
both          32434489
right_only      206209
left_only            0
Name: count, dtype: int64

Comment: This merge shows that while all rows in df_ords_prior are matched (left_only=0), there are 206,209 rows in df_ords that do not have corresponding order_id entries in df_ords_prior.

In [37]:
# Merge df_ords_prior with df_ords using the argument how = 'inner'
df_merged1 = df_ords_prior.merge(df_ords, on='order_id', how='inner', indicator=True)
# Check the value counts of the _merge column
df_merged1['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

Comment: This result reflects only the rows that match between the two datasets. It confirms that there are 32,434,489 rows with order_id in both datasets, consistent with the 'both' count from the outer join.

Overall Comment:
The two datasets have a significant overlap, with 32,434,489 matching rows (order_id values).
However, df_ords has 206,209 extra rows (indicated by right_only), which are not present in df_ords_prior. These should be reviewed for relevance, as they might represent new or unmatched data.

# Step 2

In [64]:
# Check the dimensions of df_merge
df_merge.shape

(32434489, 11)

In [66]:
# Export data df_merge dataframe to pkl
df_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))